x-pack/plugins/stack_connectors/common/inference/schema.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { schema } from '@kbn/config-schema';
+
+export const ConfigSchema = schema.object({
+  provider: schema.string(),
+  taskType: schema.string(),
+  inferenceId: schema.string(),
+  providerConfig: schema.object({}, { unknowns: 'allow', defaultValue: {} }),
+  taskTypeConfig: schema.object({}, { unknowns: 'allow', defaultValue: {} }),
+});
+
+export const SecretsSchema = schema.object({
+  providerSecrets: schema.object({}, { unknowns: 'allow', defaultValue: {} }),
```

Comment:
I don't believe we've had anyone use `unknowns: 'allow'` before - and I'm wondering if this will impact updates.  I'm not sure this is still the case, but we used to do "partial updates" of some of these things, which is why we don't allow fields to be `undefined` (`null` is fine though).  Trying to partially update an object with a field `x` set to some value, with an object that doesn't have an `x` property, would leave the `x` property in the updated object.

I'll see if I can test this myself, but would probably be good to have a function test to make sure an `update` operation on the rule works, with such a scenario.

... a bit later ...

Tested, and it appears to work in `config` and `secrets`.  I was wondering if AAD (part of encrypted SO's) would be affected, it also does not seem to have issues - to test that you have to try to execute the connector.  

I'll ask the team if anyone knows if we specifically support this now, but I'm thinking over the years we've ended up "fixing" this, and maybe not realized it :-)

Reply:
No response from the team, so I think "we're not sure" :-)

Given my tests yesterday, I think it's ok.  However, I think you should add a function test where you run the connector once, then update the config and secrets, make sure you can GET the connector back after the update, and  then try to execute the connector - it doesn't have to successfully run at the service level, we're wanting to make sure that the encrypted saved object (ESO) doesn't become corrupt because of the update - it would fail during the execute in the alerting framework code to decrypt the ESO, before it even tries to run the connector.

Reply:
I like @pmuellr's suggestion of adding a test to confirm the connector can execute after this type of update.

I believe this works because the SO update operation pulls the ESO document prior to updating the attributes, and is able to effectively perform a full update rather than a partial. And though this may work, I think it is best practice to keep a strict schema for ESOs - particularly for AAD attributes.

I just want to loop in @azasypkin here as well, as he may have some insight or opinion to share.

Reply:
💯 to the functional test.

> If the shape of an attribute is not definable with a strict schema, it may not be a good candidate for AAD.

Yes, the decision to include anything in the AAD should always be explicit, and the developer should understand why making a specific attribute part of the AAD makes sense and how this attribute relates to the encrypted value. Otherwise, including an attribute in the AAD makes it harder to reason about the ESO and could also affect the performance and maintainability of operations with such an object. Other than that, I agree with everything said above and +100 to the functional test — it will definitely pay off in the long term.

Reply:
Some context.  The object fields being added to this connector's config and secrets end up in those fields in the ESO - `secrets` is encrypted, `config` is AAD: https://github.com/elastic/kibana/blob/920d782392a4ff327fb6e59ec148c82f2b142b2a/x-pack/plugins/actions/server/saved_objects/index.ts#L72-L76

The alerting framework doesn't define the contents of these fields, they are populated by the connector type.  To make things easier (on everyone) we do not allow connector developers to specify WHAT should be in AAD, we take their entire config object and treat it as AAD.

If for some reason this is unacceptable, I think the other option is to save these as JSON-ified string values instead, which the connector would then have to serialize / parse in the connector's code.  And that would also make using with terraform/etc type tooling more cumbersome.

Reply:
Ah, I see, thanks for the context. I didn’t realize it was for the consumers of the alerting framework or second-order consumers of the ESO API. I agree, exposing the notion of AAD to these consumers or adding some additional machinery would make things more complex, confusing, and error-prone. In this case, having a more strictly defined shape for that config on the consumer side, rather than unknown, would at least provide a bit more transparency about what exactly is being used as part of the AAD, even if some of those fields might not be relevant to AAD.